pkg/crc/machine/generate_bundle.go

```diff
@@ -0,0 +1,128 @@
+	if err != nil {
+		return err
+	}
+
+	err = removePullSecret(sshRunner)
+	if err != nil {
+		return errors.Wrap(err, "Error loading bundle metadata")
+	}
+
+	// Stop the cluster
+	currentState, err := client.Stop()
+	if err != nil {
+		return err
+	}
+	if currentState != state.Stopped {
+		return fmt.Errorf("VM is not stopped, current state is %s", currentState.String())
+	}
+
+	// Create a tmp directory with name of <bundle>_custom
+	tmpBaseDir, err := ioutil.TempDir("", "crc_custom_bundle")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpBaseDir)
+
+	// Create the custom bundle directory which is used as top level directory for tarball during compression
+	currentBundleName := bundle.GetBundleNameWithoutExtension(bundleMetadata.GetBundleName())
+	customBundleName := fmt.Sprintf("%s_custom", currentBundleName)
+	customBundleDir := filepath.Join(tmpBaseDir, customBundleName)
+
+	copier, err := bundle.NewCopier(bundleMetadata, tmpBaseDir, customBundleName)
+	if err != nil {
+		return err
+	}
+
+	if err := copier.CopyKubeConfig(); err != nil {
+		return err
+	}
+
+	kubePassword, err := cluster.GetKubeadminPassword(bundleMetadata)
+	if err != nil {
+		return err
+	}
+	if err := copier.SetKubeAdminPassword(kubePassword); err != nil {
+		return err
+	}
+
+	if err := copier.CopyPrivateSSHKey(constants.GetPrivateKeyPath()); err != nil {
+		return err
+	}
+
+	if err := copier.CopyOcExecutable(); err != nil {
+		return err
+	}
+
+	// Copy disk image
+	logging.Infof("Copy the disk image to %s", customBundleDir)
+	diskPath, err := copyDiskImage(customBundleDir, bundleMetadata.GetDiskImagePath())
+	if err != nil {
+		return err
+	}
+
+	if err := copier.SetDiskImage(diskPath); err != nil {
+		return err
+	}
+
+	return copier.GenerateBundle(customBundleName)
+}
+
+func loadVM(client *client) (*bundle.CrcBundleInfo, *crcssh.Runner, error) {
+	libMachineAPIClient, cleanup := createLibMachineClient()
+	defer cleanup()
+
+	host, err := libMachineAPIClient.Load(client.name)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "Cannot load machine")
+	}
+
+	_, err = host.Driver.GetState()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "Cannot get machine state")
+	}
+
+	crcBundleMetadata, err := getBundleMetadataFromDriver(host.Driver)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "Error loading bundle metadata")
+	}
+
+	instanceIP, err := getIP(host, client.useVSock())
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "Error getting the IP")
+	}
+	sshRunner, err := crcssh.CreateRunner(instanceIP, getSSHPort(client.useVSock()), crcBundleMetadata.GetSSHKeyPath(), constants.GetPrivateKeyPath(), constants.GetRsaPrivateKeyPath())
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "Error creating the ssh client")
+	}
+
+	return crcBundleMetadata, sshRunner, nil
+}
+
+func removePullSecret(sshRunner *crcssh.Runner) error {
+	defer sshRunner.Close()
+	ocConfig := oc.UseOCWithSSH(sshRunner)
```

Comment:
@cfergeau `cluster.RemovePullSecretFromCluster` needs `ocConfig` and `sshRunner` both because using sshRunner we need to remove the pull secret from the VM and oc config for removing it from the openshift pull secret side.

Reply:
But couldn't `cluster.RemovePullSecretFromCluster` create its `ocConfig`?
```
func RemovePullSecretFromCluster(sshRunner) ... {
    ocConfig := oc.UseOCWithSSH(sshRunner)
    ...
}
```

Reply:
Ah ok, I did not check the content of the cluster package, makes sense. Having a cluster object containing the sshrunner/occonfig/... would probably be useful rather than passing these as args to each method.